### ISSUE TITLE
[QUESTION] Same global_batch_size but a smaller micro_batch_size is faster?

### ISSUE THREAD

**Issue by SefaZeng** [NVIDIA/Megatron-LM#509]
**Your question**
I try to set different micro_batch_size with the same global_batch_size to find a max micro_batch_size fits the GPU memory. But I find when the micro_batch_size goes smaller the training time for one iteration is faster. Is this correct? Or did I get something wrong here?
As for my experience, a larger batch size in each GPU will be faster as it will decrease the time for gradient accumulation. 
Specifical data for the batch_size and time:
```
global_batch_size=256

micro_batch_size=16
steps: 2 iter time (s): 2.261
----------------------
micro_batch_size=8
steps: 2 iter time (s): 1.412
----------------------
micro_batch_size=4
steps: 2 iter time (s): 0.989

```

**Comment by deepakn94** [comment]
This does not seem right. Your intuition is right: larger microbatch sizes should be slightly faster. Increasing batch size keeping the DP size the same should also increase throughput (samples / second) since data-parallel communication is less frequent.

**Comment by SefaZeng** [comment]
> This does not seem right. Your intuition is right: larger microbatch sizes should be slightly faster. Increasing batch size keeping the DP size the same should also increase throughput (samples / second) since data-parallel communication is less frequent.

I checked this issue in both the 7b model and the 600M model, it seems true for both settings.  The scripts is like:
```
TP=2
PP=2
ZERO_STAGE=0
 
GPUS_PER_NODE=8

# 7 B   
HIDDEN_SIZE=4096 # e.g. llama-13b: 5120
FFN_HIDDEN_SIZE=11008 # e.g. llama-13b: 13824
NUM_LAYERS=32 # e.g. llama-13b: 40
NUM_HEADS=32 # e.g. llama-13b: 40
SEQ_LENGTH=4096
NUM_KV_HEADS=32 # llama2 70B uses GQA
        
MICRO_BATCH_SIZE=1
GLOBAL_BATCH_SIZE=1024 # e.g. llama: 4M tokens
TRAIN_STEPS=250000 # e.g. llama: 1T tokens / 4M tokens_per_batch = 250000 steps
LR=3e-4 
MIN_LR=3e-5
LR_WARMUP_STEPS=2000
WEIGHT_DECAY=0.1
GRAD_CLIP=1

ds_args=""
ds_args=" --deepspeed ${ds_args}"
ds_args=" --deepspeed_config=$DS_CONFIG ${ds_args}"
ds_args=" --zero-stage=$ZERO_STAGE ${ds_args}"
ds_args=" --deepspeed-activation-checkpointing ${ds_args}"
 
 
DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT"
 
#$DISTRIBUTED_ARGS
torchrun $DISTRIBUTED_ARGS \
       ${CODE_PATH}/pretrain_gpt.py \
       --tensor-model-parallel-size $TP \
       --pipeline-model-parallel-size $PP \
       --num-layers $NUM_LAYERS \
       --hidden-size $HIDDEN_SIZE \
       --ffn-hidden-size $FFN_HIDDEN_SIZE \
       --num-attention-heads $NUM_HEADS \
       --micro-batch-size $MICRO_BATCH_SIZE \
       --global-batch-size $GLOBAL_BATCH_SIZE \
       --seq-length $SEQ_LENGTH \
       --max-position-embeddings $SEQ_LENGTH \
       --train-iters $TRAIN_STEPS \
       --save $CHECKPOINT_PATH \
       --load $CHECKPOINT_PATH \
       --train-data-path $DATASET \
       --valid-data-path $DATASET_VALID \
       --test-data-path $DATASET_VALID \
       --data-impl mmap \
       --tokenizer-type MT5Tokenizer \
       --tokenizer-model $TOKENIZER_PATH \
       --distributed-backend nccl \
       --lr $LR \
       --lr-decay-style cosine \
       --min-lr $MIN_LR \
       --weight-decay $WEIGHT_DECAY \
       --clip-grad $GRAD_CLIP \
       --lr-warmup-iters $LR_WARMUP_STEPS \
       --optimizer adam \
       --adam-beta1 0.9 \
       --adam-beta2 0.95 \                                                                                                                            
       --log-interval 1 \
       --save-interval 10000 \
       --eval-interval 1000 \
       --eval-iters 1000 \
       --fp16 \
       --no-query-key-layer-scaling \
       --attention-dropout 0 \
       --hidden-dropout 0 \
       --disable-bias-linear \
       $ds_args
```

**Comment by SefaZeng** [comment]
I found the MICRO_BATCH_SIZE and GLOBAL_BATCH_SIZE in the deepspeed_config.json is not the same, which is the reason for this issue...
